sway-core/src/semantic_analysis/namespace/root.rs

```diff
@@ -47,26 +105,149 @@ impl Root {
+        engines: &Engines,
+        symbol: &Ident,
+        decl: ty::TyDecl,
+    ) -> Result<ty::TyDecl, ErrorEmitted> {
+        let type_info = match decl.clone() {
+            ty::TyDecl::StructDecl(struct_decl) => TypeInfo::Struct(DeclRef::new(
+                struct_decl.name.clone(),
+                struct_decl.decl_id,
+                struct_decl.name.span(),
+            )),
+            ty::TyDecl::EnumDecl(enum_decl) => TypeInfo::Enum(DeclRef::new(
+                enum_decl.name.clone(),
+                enum_decl.decl_id,
+                enum_decl.name.span(),
+            )),
+            ty::TyDecl::TypeDecl(type_decl) => {
+                let type_decl = engines.de().get_type(&type_decl.decl_id);
+                engines.te().get(type_decl.ty.unwrap().type_id)
+            }
+            _ => {
+                return Err(handler.emit_err(CompileError::SymbolNotFound {
+                    name: symbol.clone(),
+                    span: symbol.span(),
+                }))
+            }
+        };
+
+        self.resolve_associated_type_from_type_id(
+            handler,
+            engines,
+            symbol,
+            engines.te().insert(engines, type_info),
+        )
+    }
+
+    fn resolve_associated_type_from_type_id(
+        &self,
+        handler: &Handler,
+        engines: &Engines,
+        symbol: &Ident,
+        type_id: TypeId,
+    ) -> Result<ty::TyDecl, ErrorEmitted> {
+        for trait_item in self.implemented_traits.get_items_for_type(engines, type_id) {
+            match trait_item {
+                ty::TyTraitItem::Fn(_) => {}
+                ty::TyTraitItem::Constant(_) => {}
+                ty::TyTraitItem::Type(type_ref) => {
+                    let type_decl = engines.de().get_type(type_ref.id());
+                    if type_decl.name.as_str() == symbol.as_str() {
+                        return Ok(ty::TyDecl::TypeDecl(TypeDecl {
+                            name: type_decl.name.clone(),
+                            decl_id: *type_ref.id(),
+                            decl_span: type_decl.name.span(),
+                        }));
+                    }
+                }
+            }
+        }
+
+        Err(handler.emit_err(CompileError::SymbolNotFound {
+            name: symbol.clone(),
+            span: symbol.span(),
+        }))
+    }
+
+    fn resolve_symbol_helper(
+        &self,
+        handler: &Handler,
+        engines: &Engines,
+        mod_path: &Path,
+        symbol: &Ident,
+        module: &Module,
+    ) -> Result<ty::TyDecl, ErrorEmitted> {
+        let true_symbol = self[mod_path]
+            .use_aliases
+            .get(symbol.as_str())
+            .unwrap_or(symbol);
+        match module.use_synonyms.get(symbol) {
+            Some((_, _, decl @ ty::TyDecl::EnumVariantDecl { .. }, _)) => Ok(decl.clone()),
+            Some((src_path, _, _, _)) if mod_path != src_path => {
+                // TODO: check that the symbol import is public?
```

Comment:
Is this still TODO?

Reply:
If you think this is in scope for this PR we should do it, otherwise I think it's fine to leave as is.

Reply:
I created an issue for this to be fixed in another PR.